Added SyncFail hooks

simster7 opened the pull request:
Added SyncFail hooks from: https://github.com/argoproj/argo-cd/issues/815 as well as some fixture and helper function changes

simster7 commented on controller/sync.go:
```diff
@@ -241,14 +240,17 @@ func (sc *syncContext) sync() {
 
 	// any running tasks, lets wait...
-	if tasks.Find(func(t *syncTask) bool { return t.running() }) != nil {
+	if tasks.Any(func(t *syncTask) bool { return t.running() }) {
```
A helper function borrowed from Pythons `any()` that I believe is more explicit than `task.Find`

simster7 commented on test/e2e/fixture/app/actions.go:
```diff
@@ -25,6 +25,11 @@ func (a *Actions) DeleteFile(file string) *Actions {
 	return a
 }
 
+func (a *Actions) AddFile(fileName, fileContents string) *Actions {
```
The SyncFail Hook test required two hooks. Instead of adding one to the `testdata/hook` folder, I added this fixture to add files on the fly.

alexec replied:
Good stuff.

alexec commented on controller/sync.go:
```diff
@@ -270,23 +272,43 @@ func (sc *syncContext) sync() {
+		// If tasks are already completed, don't run them again
+		if syncFailTasks.All(func(task *syncTask) bool { return task.completed() }) {
+			if syncFailTasks.All(func(task *syncTask) bool { return task.successful() }) {
+				sc.setOperationPhase(v1alpha1.OperationFailed, fmt.Sprintf("%s; SyncFail hooks applied successfully", failMessage))
```
I think we should have a resource result here. I don't think the operation failed because the failure hooks failed.

simster7 replied:
The operation failed because the failure hooks were triggered at all, so regardless of whether or not the failure hooks are successful, the overall operation should fail.

alexec replied:
ah - of course - you should probably maintain the original error message mind you - failed sync failure hooks could use the `setResult` func

alexec commented on controller/sync_tasks.go:
```diff
@@ -105,6 +105,35 @@ func (s syncTasks) Filter(predicate func(task *syncTask) bool) (tasks syncTasks)
+	return trueTasks, falseTasks
+}
+
+func (s syncTasks) All(predicate func(task *syncTask) bool) bool {
```
Test?

simster7 replied:
Added test

alexec commented on controller/sync_tasks.go:
```diff
@@ -105,6 +105,35 @@ func (s syncTasks) Filter(predicate func(task *syncTask) bool) (tasks syncTasks)
+func (s syncTasks) Split(predicate func(task *syncTask) bool) (trueTasks, falseTasks syncTasks) {
+	for _, task := range s {
+		if predicate(task) {
+			trueTasks = append(trueTasks, task)
```
Untested?

simster7 replied:
Added test

alexec commented on test/e2e/fixture/fixture.go:
```diff
@@ -271,6 +271,19 @@ func Delete(path string) {
+	CheckError(ioutil.WriteFile(filepath.Join(repoDirectory(), path), []byte(contents), 0644))
+
+	FailOnErr(Run(repoDirectory(), "git", "diff"))
+	FailOnErr(Run(repoDirectory(), "git", "add", "*"))
```
should be dot rather than star?

alexec commented:
I'd like @jessesuen 's thoughts on this.

alexec commented on controller/sync.go:
```diff
@@ -270,23 +272,43 @@ func (sc *syncContext) sync() {
+func (sc *syncContext) runFailedTasksIfAny(syncFailTasks syncTasks, failMessage string) {
+	if len(syncFailTasks) > 0 {
+		// If tasks are already completed, don't run them again
+		if syncFailTasks.All(func(task *syncTask) bool { return task.completed() }) {
```
No test coverage?

alexec replied:
Note - you only get test coverage from tests in the same directory.

simster7 replied:
Added tests in the same directory! https://github.com/argoproj/argo-cd/pull/1795/files#diff-eb441ec1bfbc70512fca4942a3d0915eR493

simster7 replied:
https://github.com/argoproj/argo-cd/blob/54e6d2efcc983ae372a43d0b95b1984feeb24838/controller/sync_test.go#L493-L569

simster7 commented:
@alexec Added more test coverage and addressed your comment here: https://github.com/argoproj/argo-cd/pull/1795#discussion_r295991147

simster7 commented on controller/sync_tasks.go:
```diff
@@ -122,6 +151,12 @@ func (s syncTasks) String() string {
 	return "[" + strings.Join(values, ", ") + "]"
 }
 
+func (s syncTasks) ApplyResourceResult(sc *syncContext, syncStatus v1alpha1.ResultCode, operationState v1alpha1.OperationPhase, message string) {
```
Another nifty helper function!

alexec replied:
unused?

simster7 commented:
@alexec Should be good. Failed test seems unrelated.

alexec commented on controller/sync.go:
```diff
@@ -277,23 +279,49 @@ func (sc *syncContext) sync() {
+			sc.setOperationPhase(v1alpha1.OperationFailed, failMessage)
+		}
+	} else {
+		sc.setOperationPhase(v1alpha1.OperationFailed, failMessage)
```
I think you should do this at the start of the function.

alexec replied:
I don't think that failing to run any of the tasks (including panic) should affect this status.

simster7 replied:
This actually can't be at the beginning of the function. The operation is technically still `OperationRunning` until all of the `SyncFail` Hooks are done. Since this function triggers the `SyncFail` hooks to begin with, setting the operation to `OperationFailed` will prematurely end it and the fail hooks will not be run

alexec commented on controller/sync.go:
```diff
@@ -277,23 +279,49 @@ func (sc *syncContext) sync() {
 	}
 }
 
+func (sc *syncContext) runFailedTasksIfAny(syncFailTasks syncTasks, failMessage string) {
```
I think we need a new name for this. Maybe `setOperationFailed`?

simster7 replied:
Fixed

alexec commented on controller/sync.go:
```diff
@@ -277,23 +279,49 @@ func (sc *syncContext) sync() {
+					return task.successful()
+				})
+				successfulTasks.ApplyResourceResult(sc, v1alpha1.ResultCodeSynced, v1alpha1.OperationSucceeded, "applied successfully")
+				failedTasdks.ApplyResourceResult(sc, v1alpha1.ResultCodeSynced, v1alpha1.OperationFailed, "failed to apply")
```
typo?

simster7 replied:
Fixed

alexec commented on controller/sync.go:
```diff
@@ -277,23 +279,49 @@ func (sc *syncContext) sync() {
+		if syncFailTasks.All(func(task *syncTask) bool { return task.completed() }) {
+			if syncFailTasks.All(func(task *syncTask) bool { return task.successful() }) {
+				sc.setOperationPhase(v1alpha1.OperationFailed, failMessage)
+				syncFailTasks.ApplyResourceResult(sc, v1alpha1.ResultCodeSynced, v1alpha1.OperationSucceeded, "applied successfully")
```
As discussed, `runTasks` will store the result.

simster7 replied:
I'm almost certain that this particular resource set is actually necessary. Will review tomorrow morning to make sure and get back to you.

simster7 replied:
Turns out they weren't necessary, I removed them.

simster7 commented:
@alexec Should be good, addressed your comments and added an additional test. The failing e2e test (`TestKustomize2AppSource`) on circleci seems flakey. It passes locally for me.

alexec commented on test/e2e/fixture/app/expectation.go:
```diff
@@ -98,6 +98,7 @@ func ResourceResultNumbering(num int) Expectation {
 func ResourceResultIs(result ResourceResult) Expectation {
 	return func(c *Consequences) (state, string) {
 		for _, res := range c.app().Status.OperationState.SyncResult.Resources {
+			fmt.Println("TRYING: ", res)
```
delete?

simster7 replied:
Yup, my bad

alexec commented:
I've had a completely mad idea. Let me try it out and get back to you.

alexec commented:
Ok. So I poked about to understand this and it made me take pause for a moment. 

It's worthwhile bearing in mind that a sync op can pass or fail, but the underlying resources are degraded. Specifically, the health of the last phase/wave is ignored. 

The question, we come back to, is when do you need this triggered?

simster7 commented:
Not too sure what you mean by "this" in "when do you need 'this' triggered?", but if you mean the SyncFail hooks I think it would make sense to trigger them anytime there is a failure in the sync operation, regardless of where it occurs. This would be any time the operation phase is set to `OperationFailed`. I don't think this should run if a sync operation is successful and then only afterwards the relevant resources become degraded.

alexec commented on controller/sync.go:
```diff
@@ -205,7 +205,6 @@ func (sc *syncContext) sync() {
 	if !sc.started() {
 		sc.log.Debug("dry-run")
 		if !sc.runTasks(tasks, true) {
-			sc.setOperationPhase(v1alpha1.OperationFailed, "one or more objects failed to apply (dry run)")
```
why delete?

simster7 replied:
Oversight from testing.

alexec commented:
So I pull your code locally and riffed on it in my own PR to understand how it worked. 

https://github.com/argoproj/argo-cd/pull/1819

I could of points, I think we can (and probably should) implement this without the `Filter` and `Apply*` methods. You can request the tasks when you need them in `sc.setOperationFailed`.  That cuts 200 lines of code.  All other things equal, less code = less bugs.

I'm being quite opinionated about this, but you can tell that I'm pretty familiar with this piece of code.

The key thing we need to get correct is `sc.setOperationFailed`. There are key assumptions in this method. Essentially, I don't think the "if" statement is needed. I'll try and send you the changes shortly.

simster7 commented:
Sure, let me know when your changes are ready.

alexec commented:
I think we should pair. I'll OOO today, maybe tomorrow.

simster7 commented:
Sounds good, I'll stop by tomorrow

alexec commented on controller/sync.go:
```diff
@@ -277,23 +279,39 @@ func (sc *syncContext) sync() {
 	}
 }
 
+func (sc *syncContext) setOperationFailed(syncFailTasks syncTasks, failMessage string) {
```
can we rename just `message`

alexec commented on controller/sync_test.go:
```diff
@@ -518,6 +518,89 @@ func TestPersistRevisionHistoryRollback(t *testing.T) {
+	syncCtx.syncOp.SyncStrategy.Apply = nil
+	pod := test.NewPod()
+	pod.SetName("")
+	pod.SetAnnotations(map[string]string{common.AnnotationKeyHook: "PostSync"})
```
test.NewHook()

alexec replied:
```
func TestSyncFailureHookWithSuccessfulSync(t *testing.T) {
	syncCtx := newTestSyncCtx()
	syncCtx.syncOp.SyncStrategy.Apply = nil
	syncCtx.compareResult = &comparisonResult{
		managedResources: []managedResource{{Target: test.NewPod()}},
		hooks:            []*unstructured.Unstructured{test.NewHook(HookTypeSyncFail)},
	}

	syncCtx.sync()

	assert.Equal(t, OperationSucceeded, syncCtx.opState.Phase)
	// only one result, we did not run the failure failureHook
	assert.Len(t, syncCtx.syncRes.Resources, 1)
}

func TestSyncFailureHookWithFailedSync(t *testing.T) {
	syncCtx := newTestSyncCtx()
	syncCtx.syncOp.SyncStrategy.Apply = nil
	pod := test.NewPod()
	syncCtx.compareResult = &comparisonResult{
		managedResources: []managedResource{{Target: pod}},
		hooks:            []*unstructured.Unstructured{test.NewHook(HookTypeSyncFail)},
	}
	syncCtx.kubectl = &kubetest.MockKubectlCmd{
		Commands: map[string]kubetest.KubectlOutput{pod.GetName(): {Err: fmt.Errorf("")}},
	}

	syncCtx.sync()
	syncCtx.sync()

	assert.Equal(t, OperationFailed, syncCtx.opState.Phase)
	assert.Len(t, syncCtx.syncRes.Resources, 2)
}
```

alexec commented on controller/sync_tasks.go:
```diff
@@ -11,6 +11,7 @@ var syncPhaseOrder = map[v1alpha1.SyncPhase]int{
 	v1alpha1.SyncPhasePreSync:  -1,
 	v1alpha1.SyncPhaseSync:     0,
 	v1alpha1.SyncPhasePostSync: 1,
+	v1alpha1.SyncPhaseSyncFail: 2,
```
maybe test

alexec commented on controller/sync_test.go:
```diff
@@ -518,6 +518,89 @@ func TestPersistRevisionHistoryRollback(t *testing.T) {
+	assert.Equal(t, syncFailTasks[0].syncStatus, ResultCodeSynced)
+}
+
+func TestRunSyncFailHooksFailed(t *testing.T) {
```
comment explaining scenarios

alexec commented on test/e2e/hook_test.go:
```diff
@@ -116,6 +116,84 @@ func TestPostSyncHookPodFailure(t *testing.T) {
 }
 
+func TestSyncFailHookPodFailure(t *testing.T) {
+	hook2Contents := `
```
inline?

alexec commented on test/e2e/hook_test.go:
```diff
@@ -116,6 +116,84 @@ func TestPostSyncHookPodFailure(t *testing.T) {
+	Given(t).
+		Path("hook").
+		When().
+		AddFile("hook2.yaml", hook2Contents).
```
hook2 -> failure-hook

alexec commented on test/e2e/hook_test.go:
```diff
@@ -116,6 +116,84 @@ func TestPostSyncHookPodFailure(t *testing.T) {
 		Expect(NotPod(func(p v1.Pod) bool { return p.Name == "hook" }))
 }
 
+func TestSyncFailHookPodFailure(t *testing.T) {
```
add commentary

alexec commented on test/e2e/hook_test.go:
```diff
@@ -116,6 +116,84 @@ func TestPostSyncHookPodFailure(t *testing.T) {
+	Given(t).
+		Path("hook").
+		When().
+		AddFile("hook2.yaml", hook2Contents).
```
inline and name resources

alexec commented on util/hook/hook.go:
```diff
@@ -34,7 +34,7 @@ func Types(obj *unstructured.Unstructured) []v1alpha1.HookType {
 	for _, hookType := range types(obj) {
 		switch v1alpha1.HookType(hookType) {
-		case v1alpha1.HookTypePreSync, v1alpha1.HookTypeSync, v1alpha1.HookTypePostSync:
+		case v1alpha1.HookTypePreSync, v1alpha1.HookTypeSync, v1alpha1.HookTypePostSync, v1alpha1.HookTypeSyncFail:
```
test?

alexec commented on controller/sync.go:
```diff
@@ -205,7 +205,6 @@ func (sc *syncContext) sync() {
 	if !sc.started() {
 		sc.log.Debug("dry-run")
 		if !sc.runTasks(tasks, true) {
-			sc.setOperationPhase(v1alpha1.OperationFailed, "one or more objects failed to apply (dry run)")
```
still not fixed?

simster7 replied:
How did this come back?!

alexec commented on controller/sync_test.go:
```diff
@@ -518,6 +518,78 @@ func TestPersistRevisionHistoryRollback(t *testing.T) {
+	// Operation as a whole should fail
+	assert.Equal(t, OperationFailed, syncCtx.opState.Phase)
+	// failedSyncFailHook should fail
+	assert.Equal(t, syncCtx.syncRes.Resources[1].HookPhase, OperationFailed)
```
arg order?

simster7 replied:
Yup

alexec commented on test/e2e/fixture/fixture.go:
```diff
@@ -307,6 +307,19 @@ func Delete(path string) {
 	FailOnErr(Run(repoDirectory(), "git", "commit", "-am", "delete"))
 }
 
+func AddFile(path, contents string) {
```
neat

alexec commented on test/e2e/hook_test.go:
```diff
@@ -116,6 +116,83 @@ func TestPostSyncHookPodFailure(t *testing.T) {
+metadata:
+  annotations:
+    argocd.argoproj.io/hook: SyncFail
+  name: hook2
```
can we give it another name?

alexec commented on test/e2e/hook_test.go:
```diff
@@ -116,6 +116,83 @@ func TestPostSyncHookPodFailure(t *testing.T) {
+metadata:
+  annotations:
+    argocd.argoproj.io/hook: SyncFail
+  name: hook2
```
clearer name?

alexec commented on test/e2e/hook_test.go:
```diff
@@ -116,6 +116,83 @@ func TestPostSyncHookPodFailure(t *testing.T) {
+metadata:
+  annotations:
+    argocd.argoproj.io/hook: SyncFail
+  name: hook3
```
clearer name?
